CppImport/src/macro/ClangHelper.cpp

```
@@ -0,0 +1,126 @@
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+
+#include "ClangHelper.h"
+
+namespace CppImport {
+
+void ClangHelper::setSourceManager(const clang::SourceManager* sourceManager)
+{
+	sourceManager_ = sourceManager;
+}
+
+void ClangHelper::setPreprocessor(const clang::Preprocessor* preprocessor)
+{
+	preprocessor_ = preprocessor;
+}
+
+QString ClangHelper::spelling(clang::SourceLocation loc)
+{
+	return spelling(loc, loc);
+}
+
+QString ClangHelper::spelling(clang::SourceRange range)
+{
+	return spelling(range.getBegin(), range.getEnd());
+}
+
+QString ClangHelper::spelling(clang::SourceLocation start, clang::SourceLocation end)
+{
+	clang::SourceLocation b = sourceManager_->getSpellingLoc(start);
+	clang::SourceLocation e = clang::Lexer::getLocForEndOfToken(sourceManager_->getSpellingLoc(end), 0, *sourceManager_,
+																					preprocessor_->getLangOpts());
+
+	auto length = sourceManager_->getCharacterData(e) - sourceManager_->getCharacterData(b);
+	if (length > 1000000) return "ERROR_IN_GET_SPELLING";
+
+	try
+	{
+		return 0 < length ? QString::fromStdString(std::string(sourceManager_->getCharacterData(b), length)) : "";
+	}
+	catch (...)
+	{
+		return "ERROR_IN_GET_SPELLING";
```

ugh. catch ALL exceptions and return a string ? really? what if there was an out of memory exception? Surely there is a more limited set of exceptions that you want to silence here. But I'm not fond of this error mechanism. We should talk about it.